File: app/src/main/java/org/singularux/contacts/ui/ContactListViewModel.java
package org.singularux.contacts.ui;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.LiveDataReactiveStreams;
import androidx.lifecycle.ViewModel;

import org.singularux.contacts.core.ContactsPermission;
import org.singularux.contacts.core.ContactsPermissionManager;
import org.singularux.contacts.data.ContactBriefEntity;
import org.singularux.contacts.domain.ListenContactListUseCase;

import java.util.List;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;
import lombok.Getter;

@HiltViewModel
public class ContactListViewModel extends ViewModel {

    private final @Getter String[] readContactsPermissions;
    private final @Getter LiveData<List<ComponentData>> contactListLiveData;

    @Inject
    public ContactListViewModel(ContactsPermissionManager contactsPermissionManager,
                                ListenContactListUseCase listenContactListUseCase
    ) {
        this.readContactsPermissions = contactsPermissionManager.getPermissionStrings(
                ContactsPermission.READ_CONTACTS, ContactsPermission.READ_PROFILE);
        this.contactListLiveData = LiveDataReactiveStreams
                .fromPublisher(listenContactListUseCase.get());
    }

}

File: app/src/main/kotlin/org/singularux/contacts/feature/contactlist/viewmodel/ContactListViewModel.kt
package org.singularux.contacts.feature.contactlist.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.withContext
import org.singularux.contacts.feature.contactlist.domain.ListenContactListUseCase
import org.singularux.contacts.feature.contactlist.model.ContactItem
import javax.inject.Inject

@HiltViewModel
class ContactListViewModel @Inject constructor(
    private val listenContactListUseCase: ListenContactListUseCase
) : ViewModel() {

    val contactListFlow = listenContactListUseCase()
        .map { contactBriefEntityList ->
            withContext(Dispatchers.Default) {
                contactBriefEntityList.map { contactBriefEntity ->
                    ContactItem(
                        lookupKey = contactBriefEntity.lookupKey,
                        displayName = contactBriefEntity.displayName,
                        thumbnailUri = contactBriefEntity.thumbnailUri
                    )
                }
            }
        }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

}